In [54]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import r2_score

## Part 1. Loading the dataset

In [55]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

url = "https://raw.githubusercontent.com/COMPSCI-383-Fall2024/assignment-4-equation-of-a-slime-<redacted>/refs/heads/main/science_data_large.csv?token=GHSAT0AAAAAACXIVX5R3S77TCZ6YZB2A5HGZYQP74Q"
data = pd.read_csv(url)

print("First 15 rows of the dataset:")
print(data.head(15))

print("\nSummary of the dataset:")
print(data.info())

First 15 rows of the dataset:
    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

Summary of the dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL    

## Part 2. Splitting the dataset

In [56]:
# Take the pandas dataset and split it into our features (X) and label (y)

X = data[['Temperature °C', 'Mols KCL']]
y = data['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)


## Part 3. Perform a Linear Regression

In [57]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_data = [[420, 365]]
predicted_output = model.predict(sample_data)
print(f"Predicted output for the sample data {sample_data}: {predicted_output[0]}")

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
model_score = model.score(X_test, y_test)
print(f"Model R^2 score: {model_score}")

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_
print(f"Coefficients: {coefficients}")
print(f"Intercept: {intercept}")

Predicted output for the sample data [[420, 365]]: 331323.71330011485
Model R^2 score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


The score reported by the LinearRegression model is the R^2 score, which is the coefficient of determination. This score measures how well the model explains the variance in the target variable, which is: Size nm^3. An R^2 score of 1 indicates a perfect fit, meaning the model explains 100% of the variance in the data. A score of 0 means the model does not explain any variance beyond what the mean of the target would explain. For our model, a higher R² score means the features (Temperature °C, Mols KCL nm^3) can accurately predict the target (Size nm^3).



Let $T$ be Temperature °C, and let $M$ be Mols KCL nm^3

Answer: $h(x) = -409391.480 + 866.146(T) + 1032.695(M)$

## Part 4. Use Cross Validation

In [58]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
cross_val_scores = cross_val_score(model, X, y, cv=10, scoring='r2')
# Using a (90% train, 10% test) split in each fold
# On each iteration, 1 fold is test and remaining 9 folds are for training the model

# Report on their finding and their significance
print(f"Cross-validation R² scores: {cross_val_scores}") # Shows the R^2 values for each fold
print(f"Mean R² score: {cross_val_scores.mean()}") # Displays the average R^2 score from all folds, indicating overall model performance
print(f"Standard deviation of R² scores: {cross_val_scores.std()}") # Indicates the variability in model performance across different folds
print(f"Minimum R² score: {cross_val_scores.min()}") # Identifies the lowest R^2 score, revealing the worst performance among the folds

Cross-validation R² scores: [0.81123596 0.86440978 0.87808742 0.86561069 0.87495621 0.84484397
 0.87941022 0.86349411 0.78353682 0.88686516]
Mean R² score: 0.8552450341984701
Standard deviation of R² scores: 0.03152876296534241
Minimum R² score: 0.7835368198922626


## Part 5. Using Polynomial Regression

In [59]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

# Do 90% train, 10% split again
X_train, X_test, y_train, y_test = train_test_split(X_poly, y, test_size=0.1, random_state=42)

model = LinearRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
r2 = r2_score(y_test, y_pred)
coefficients = model.coef_
intercept = model.intercept_

# Report on the metrics and output the resultant equation as you did in Part 3.
print(f"Polynomial Regression R² score: {r2}")
print(f"Coefficients: {coefficients}")
print(f"Intercept: {intercept}")

Polynomial Regression R² score: 1.0
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 2.0477105863392353e-05


The polynomial regression model achieved an R^2 score of 1.0, meaning that it perfectly explains the variance in the target variable. This suggests that the model captures all underlying patterns in the data, and all the data points are explained accurately. This model is further correct because when we switch to a degree of 1 (not shown in code), the output matches the linear regression model, which is accurate.


Let the coefficients be defined as: $x_1, x_2, x_3, x_4, x_5, x_6$

Answer: $h(x) = 2.048 \times 10^{-5} + 12(x_2) - 1.272 \times 10^{-7}(x_3) + 1.265 \times 10^{-11}(x_4) + 2(x_5) + 2.857 \times 10^{-2}(x_6)$
